README.md
# report

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2018 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-02-02
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-02-02
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/report/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/report/compare/0.1.0...0.1.1


test/report/core_test.clj
(ns report.core-test
  (:require [clojure.test :refer :all]
            [report.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject report "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}

  :dependencies [[org.clojure/clojure "1.9.0"]
                 [org.postgresql/postgresql "42.1.4.jre7"]
                 [org.clojure/java.jdbc "0.7.4"]
                 [org.clojure/data.csv "0.1.4"]

                 [ring "1.6.3"]
                 [ring-cors "0.1.11"]
                 [ring/ring-core "1.6.3"]
                 [ring/ring-mock "0.3.2"]
                 [ring/ring-json "0.5.0-beta1"]
                 [ring/ring-jetty-adapter "1.6.3"]

                 [csv-map "0.1.2"]
                 [clj-x256 "0.0.1"]
                 [clj-http "3.7.0"]
                 [clj-time "0.14.2"]
                 [environ "1.1.0"]

                 [compojure "1.6.0"]
                 [cheshire "5.8.0"]
                 [reloaded.repl "0.2.4"]
                 [com.stuartsierra/component "0.3.2"]
                 [javax.servlet/servlet-api "3.0-alpha-1"]]

  :min-lein-version "2.0.0"

  :plugins [[lein-ring "0.12.2"]
            [lein-heroku "0.5.3"]
            [lein-environ "1.1.0"]
            [environ/environ.lein "0.3.1"]]

  :hooks [environ.leiningen.hooks]

  :ring {:handler report.core/app-handler
         :auto-reload? true
         :auto-refresh? false}

  :profiles {:dev {:source-paths ["dev"]
                   :dependencies [[org.clojure/java.classpath "0.2.3"]
                                  [org.clojure/tools.namespace "0.3.0-alpha4"]]
                   }
             :uberjar {:aot
                       :all
                       }
             :production {:env {:production true}}
             }



  )


dev/user.clj
(ns user

  (:require [report.core :as rc]
            [report.system :as rs]
            [clojure.repl :refer :all]
            [environ.core :refer [env]]
            [clojure.tools.namespace.repl :refer (refresh refresh-all)]
            )

  )

(def system nil)

(defn init
  "Constructs the current development system."
  []
  (let [port (Integer. (env :port "5000"))]

    (alter-var-root #'system
                    (constantly (rs/system rc/myroutes port )))

    )
  )

(defn start []
  "Starts the current development system."
  (alter-var-root #'system rs/start)
  )

(defn stop
  "Shuts down and destroys the current development system."
  []
  (alter-var-root #'system
                  (fn [s] (when s (rs/stop s)))))

(defn go
  "Initializes the current development system and starts it running."
  []
  (init)
  (start))

(defn reset []
  (stop)
  (refresh :after 'user/go))

src/report/system.clj
(ns report.system

  (:use [compojure.core :refer :all])

  (:require [cheshire.core :refer :all]
            [environ.core :refer [env]]
            [ring.adapter.jetty :as jetty]
            [ring.middleware.cors :refer [wrap-cors]]
            [ring.middleware.multipart-params :refer [wrap-multipart-params]]
            )

  )

;-----------------------------------------------------------------------------------------*
;this section content a construction function for the application                         *
;It specify configuration of parameters and will create instances and return a data       *
;as a map                                                                                 *
;-----------------------------------------------------------------------------------------*

(defn system[route port]
  "Returns a new instance of the whole application."

  {:db        (System/getenv "DATABASE_URL")

   :handler   {:status  200
               :headers {"Content-Type"                     "application/json"
                         "Access-Control-Allow-Origin"      "*"
                         "Access-Control-Allow-Credentials" "false"
                         "Access-Control-Allow-Methods"     "POST, GET, OPTIONS"
                         "Access-Control-Allow-Headers"     "Accept, Content-Type"}
               }

   :route     route

   :webserver nil

   :port      port
   }
  )

;-----------------------------------------------------------------------------------------*
;this section content functions that start and stop the web server                        *
;in that they return a new value representing the "started" or "stopped" system,          *
;but they also have to perform side effects along the way, such as opening a connection   *
;to a database or starting a web server.                                                  *
;-----------------------------------------------------------------------------------------*

(defn start [system]
  "Performs side effects to initialize the system, acquire resources,
  and start it running. Returns an updated instance of the system."
  (let [port (Integer. (env :port "5000"))
        server (jetty/run-jetty (wrap-cors (wrap-multipart-params (:route system))
                                           :access-control-allow-methods [:get :post :delete :options]
                                           :access-control-allow-headers ["Accept, Content-Type"]
                                           :access-control-allow-origin [#"http://localhost:4200"]
                                           )
                                {:port port :join? false})]
    (assoc system :webserver server))

  )
(defn stop [system]
  "Performs side effects to shut down the system and release its
  resources. Returns an updated instance of the system."

  (when (:webserver system)

    (.stop (:webserver system))

    (assoc system :webserver nil)
    )

  )

;-----------------------------------------------------------------------------------------*


src/report/core.clj
(ns report.core

  (:use [clojure.java.io]
        [clojure.set]
        [compojure.core :refer :all])

  (:require [clojure.java.io]
            [clojure.data.csv]
            [cheshire.core :refer :all]
            [compojure.core :refer :all]
            [compojure.route :as route]

            [clojure.java.jdbc :as cj]
            [clj-time.coerce :as clt]
            [environ.core :refer [env]]

            [ring.adapter.jetty :as jetty]
            [ring.util.response :refer [response]]
            [ring.middleware.cors :refer [wrap-cors]]
            [ring.middleware.params :refer [wrap-params]]
            [ring.middleware.json :refer [wrap-json-response]]
            [ring.middleware.multipart-params :refer [wrap-multipart-params]])

  )

;-----------------------------------------------------------------------------------------*

;convert string to integer
(defn to-int [s]
  (if (integer? s)
    s
    (try
      (Integer. s)
      (catch Exception e nil))
    )
  )

;connect to the database
(def db (System/getenv "DATABASE_URL"))
(def db-url (env :database-url))

;-----------------------------------------------------------------------------------------*
;this section content functions that do multiple operation on the csv file in order to    *
;get its content and generate a report                                                    *
;-----------------------------------------------------------------------------------------*

(defn read-file [file]
  (clojure.string/split-lines (slurp file)))

(defn get-org-id [file]
  (apply str (->> (clojure.string/split-lines file)
                  (map #(clojure.string/split % #","))
                  (map #(nth % 1))
                  (map #(clojure.string/split % #"/"))
                  (map #(nth % 0))
                  (map #(clojure.string/replace % "\"" ""))))
  )

(defn get-org-memory [org-name file-line]
  (->> (filter #(= org-name (get-org-id %)) file-line)
       (map #(clojure.string/split % #","))
       (map #(nth % 2))
       (map #(re-find #"(\d+)" %))
       (map second)
       (map to-int)
       (reduce +)
       (hash-map :name org-name :size))
  )

(defn get-org-and-id [file-line org-list]
  (map (fn [org-name]
         (get-org-memory org-name file-line)) org-list)
  )

(defn get-lines-report [file-lines]
  (let [org-list (distinct (map #(get-org-id %) file-lines))]
    (get-org-and-id file-lines org-list))
  )

(defn file-report [file]
  (let [file-line (read-file file)]
    (get-lines-report file-line))
  )

;-----------------------------------------------------------------------------------------*
;this section content functions that connect and upload a report into the database        *
;                                                                                         *
;-----------------------------------------------------------------------------------------*

(defn get-name-and-size [report date]
  (let [_name (:name report)
        _size (:size report)
        _date (clt/to-sql-date (clt/to-string date))]

    (cj/insert! db-url :cloudrepo_report {:organization _name
                                      :size         _size
                                      :date         _date})
    )
  )

(defn upload-report-to-database [report date]
  (map #(get-name-and-size % date) report)
  )

(defn upload-file [file]
  (let [file-name (file :filename)
        size (file :size)]
    (do
      (file-report file))
    )
  )

;-----------------------------------------------------------------------------------------*
;this section content queries functions that read the report data stored in the postgres  *
;database which returns a sub report by organization name or the data uploaded            *
;-----------------------------------------------------------------------------------------*

(defn get-full-report []
  (cj/query db-url ["SELECT DISTINCT ON (organization)\n
                     organization, size, date\n
                     FROM cloudrepo_report\n
                     ORDER BY organization"])
  )

(defn get-full-report-of-date []
  (cj/query db-url ["SELECT DISTINCT ON (date)\n
                     organization, size, date\n
                     FROM cloudrepo_report\n
                     ORDER BY date DESC"])
  )

(defn view-by-organization [org-name]

  (cj/query db-url ["SELECT
                    \"organization\", \"size\", \"date\"
                    FROM \"cloudrepo_report\"
                    WHERE (\"organization\" = ?) ORDER BY \"date\" DESC" org-name]))

(defn view-by-date [input]

  (let [_date (clt/to-sql-date (clt/to-string input))]

    (cj/query db-url ["SELECT
                      \"organization\", \"size\", \"date\"
                      FROM \"cloudrepo_report\"
                      WHERE (\"date\" = ?) ORDER BY \"size\" DESC" _date])

    )
  )

(defn delete-by-name [org-name]

  (cj/query db-url ["DELETE FROM
                     \"cloudrepo_report\"
                     WHERE (\"organization\" = ?) RETURNING *" org-name]
            )

  )

(defn delete-by-date [input]

  (let [_date (clt/to-sql-date (clt/to-string input))]

    (cj/query db-url ["DELETE FROM
                      \"cloudrepo_report\"
                      WHERE (\"date\" = ?) RETURNING *" _date])

    )
  )

(defn delete-full-report []

  (cj/query db-url ["DELETE FROM
                    \"cloudrepo_report\" RETURNING *"]
            )

  )

;-----------------------------------------------------------------------------------------*
;this section content functions for routes and route handler                              *
;REST API configurations are also set here                                                *
;-----------------------------------------------------------------------------------------*

(defn app-handler [request]
  {:status  200
   :headers {"Content-Type"                     "application/json"
             "Access-Control-Allow-Origin"      "*"
             "Access-Control-Allow-Credentials" "false"
             "Access-Control-Allow-Methods"     "POST, GET, DELETE, OPTIONS"
             "Access-Control-Allow-Headers"     "Accept, Content-Type"}

   :body    (generate-string request)}
  )

(defroutes myroutes (GET "/" []
                      (apply str "<h1>Hello Welcome! This is a report page of CloudRepo users</h1>")
                      )

           (GET "/report" []
             (generate-string (get-full-report))

             )

           (GET "/date" []
             (generate-string (get-full-report-of-date))

             )

           (GET "/date/:input" [input]
             (generate-string (view-by-date input))

             )

           (GET "/name/:input" [input]
             (generate-string (view-by-organization input))

             )

           (GET "/delete-date/:input" [input]
             (generate-string (delete-by-date input))

             )

           (GET "/delete-name/:input" [input]
             (generate-string (delete-by-name input))

             )

           (GET "/delete-all-record" []
             (generate-string (delete-full-report))

             )

           (POST "/file" {params :params
                          :as    req
                          }

                 (let [fileparam (get params "file")
                       file (:tempfile fileparam)
                       dateparam (get params "date")
                       date (clt/to-string dateparam)
                       report (file-report file)]

                    (upload-report-to-database report date)
                   )

                 )

           (route/not-found "<h1>Page not found or does not exit</h1>")

           )


;-----------------------------------------------------------------------------------------*
;this section content the main function that start the server                             *
;on local host port 5000                                                                  *
;-----------------------------------------------------------------------------------------*

(defn -main [& [port]]
  (let [port (Integer. (or port (env :port) 3000))]

    (jetty/run-jetty (wrap-cors (wrap-multipart-params myroutes)
                                :access-control-allow-methods [:get :post :delete :options]
                                :access-control-allow-headers ["Content-Type"]
                                :access-control-allow-origin [#"http://localhost:4200"]
                                )
                     {:port port :join? false}
                     )

    )

  )

;-----------------------------------------------------------------------------------------*
